pandas/core/groupby/groupby.py

```diff
@@ -3270,26 +3270,37 @@ def sample(
         2   blue  2
         0    red  0
         """
-        from pandas.core.reshape.concat import concat
-
+        size = algorithms.process_sampling_size(n, frac, replace)
         if weights is not None:
-            weights = Series(weights, index=self._selected_obj.index)
-            ws = [weights.iloc[idx] for idx in self.indices.values()]
-        else:
-            ws = [None] * self.ngroups
+            weights_arr = algorithms.preprocess_weights(
+                self._selected_obj, weights, axis=self.axis
+            )
 
-        if random_state is not None:
-            random_state = com.random_state(random_state)
+        random_state = com.random_state(random_state)
 
         group_iterator = self.grouper.get_iterator(self._selected_obj, self.axis)
-        samples = [
-            obj.sample(
-                n=n, frac=frac, replace=replace, weights=w, random_state=random_state
+
+        sampled_indices = []
+        for labels, obj in group_iterator:
+            grp_indices = self.indices[labels]
+            group_size = len(grp_indices)
+            if size is not None:
+                sample_size = size
+            else:
+                assert frac is not None
+                sample_size = round(frac * group_size)
+
+            grp_sample = algorithms.sample(
+                group_size,
+                size=sample_size,
+                replace=replace,
+                weights=None if weights is None else weights_arr[grp_indices],
+                random_state=random_state,
             )
-            for (_, obj), w in zip(group_iterator, ws)
-        ]
+            sampled_indices.append(grp_indices[grp_sample])
 
-        return concat(samples, axis=self.axis)
+        sampled_indices = np.concatenate(sampled_indices)
```

Comment:
Ends up much faster to concatenate the sampled ndarrays, then take once, instead of taking with each group and doing a full `pd.concat`

Reply:
nice!